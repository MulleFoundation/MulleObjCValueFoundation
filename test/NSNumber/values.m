#ifndef __MULLE_OBJC__
# import <Foundation/Foundation.h>

static inline char   *_NSComparisonResultUTF8String( NSComparisonResult result)
{
   return( result < 0 ? "<" : (result > 0 ? ">" : "="));
}


static inline char   *_MulleBoolUTF8String( BOOL result)
{
   return( result ? "YES" : "NO");
}


#else
# import <MulleObjCValueFoundation/MulleObjCValueFoundation.h>
#endif

#include <float.h>
#include <math.h>


// some harmless values :)
static double  values[] =
{
   0.8,
   12.2,
   7/10.0,
   1848.0,
   FLT_MIN,
   FLT_MIN-0.00001,
   FLT_MAX,
   FLT_MAX+0.00001,
   1.2e-63,
   1.2e+63
};


#define sizeof_array( array) ( sizeof( array) / sizeof( array[ 0]))

int   main( void)
{
   NSNumber             *nr1;
   NSNumber             *nr2;
   NSInteger            i;
   NSInteger            j;
   NSComparisonResult   result;
   BOOL                 flag;

   for( i = 0; i < sizeof_array( values); i++)
   {
      nr1 = [NSNumber numberWithDouble:values[ i]];
      for( j = i + 1; j < sizeof_array( values); j++)
      {
         nr2 = [NSNumber numberWithDouble:values[ j]];

         printf( "C: %g <op> %g :",
                     values[ i],
                     values[ j]);
         if( values[ i] < values[ j])
            printf( " <");
         if( values[ i] == values[ j])
            printf( " ==");
         if( values[ i] != values[ j])
            printf( " !=");
         if( values[ i] > values[ j])
            printf( " >");
         printf( "\n");

         printf( "C: %g <op> %g :",
                     values[ j],
                     values[ i]);
         if( values[ j] < values[ i])
            printf( " <");
         if( values[ j] == values[ i])
            printf( " ==");
         if( values[ j] != values[ i])
            printf( " !=");
         if( values[ j] > values[ i])
            printf( " >");
         printf( "\n");


         result = [nr1 compare:nr2];
         printf( "   [%g compare:%g] : %s\n",
                     values[ i],
                     values[ j],
                     _NSComparisonResultUTF8String( result));

         result = [nr2 compare:nr1];
         printf( "   [%g compare:%g] : %s\n",
                     values[ j],
                     values[ i],
                     _NSComparisonResultUTF8String( result));

         flag = [nr1 isEqual:nr2];
         printf( "   [%g isEqual:%g] : %s\n",
                     values[ i],
                     values[ j],
                     _MulleBoolUTF8String( flag));

         flag = [nr2 isEqual:nr1];
         printf( "   [%g isEqual:%g] : %s\n",
                     values[ j],
                     values[ i],
                     _MulleBoolUTF8String( flag));

         flag = [nr1 isEqualToNumber:nr2];
         printf( "   [%g isEqualToNumber:%g] : %s\n",
                     values[ i],
                     values[ j],
                     _MulleBoolUTF8String( flag));

         flag = [nr2 isEqualToNumber:nr1];
         printf( "   [%g isEqualToNumber:%g] : %s\n",
                     values[ j],
                     values[ i],
                     _MulleBoolUTF8String( flag));
      }
   }
   return( 0);
}
