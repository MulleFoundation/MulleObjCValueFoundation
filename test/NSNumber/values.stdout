C: 0.8 <op> 12.2 : < !=
C: 12.2 <op> 0.8 : != >
   [0.8 compare:12.2] : <
   [12.2 compare:0.8] : >
   [0.8 isEqual:12.2] : NO
   [12.2 isEqual:0.8] : NO
   [0.8 isEqualToNumber:12.2] : NO
   [12.2 isEqualToNumber:0.8] : NO
C: 0.8 <op> 0.7 : != >
C: 0.7 <op> 0.8 : < !=
   [0.8 compare:0.7] : >
   [0.7 compare:0.8] : <
   [0.8 isEqual:0.7] : NO
   [0.7 isEqual:0.8] : NO
   [0.8 isEqualToNumber:0.7] : NO
   [0.7 isEqualToNumber:0.8] : NO
C: 0.8 <op> 1848 : < !=
C: 1848 <op> 0.8 : != >
   [0.8 compare:1848] : <
   [1848 compare:0.8] : >
   [0.8 isEqual:1848] : NO
   [1848 isEqual:0.8] : NO
   [0.8 isEqualToNumber:1848] : NO
   [1848 isEqualToNumber:0.8] : NO
C: 0.8 <op> 1.17549e-38 : != >
C: 1.17549e-38 <op> 0.8 : < !=
   [0.8 compare:1.17549e-38] : >
   [1.17549e-38 compare:0.8] : <
   [0.8 isEqual:1.17549e-38] : NO
   [1.17549e-38 isEqual:0.8] : NO
   [0.8 isEqualToNumber:1.17549e-38] : NO
   [1.17549e-38 isEqualToNumber:0.8] : NO
C: 0.8 <op> -1e-05 : != >
C: -1e-05 <op> 0.8 : < !=
   [0.8 compare:-1e-05] : >
   [-1e-05 compare:0.8] : <
   [0.8 isEqual:-1e-05] : NO
   [-1e-05 isEqual:0.8] : NO
   [0.8 isEqualToNumber:-1e-05] : NO
   [-1e-05 isEqualToNumber:0.8] : NO
C: 0.8 <op> 3.40282e+38 : < !=
C: 3.40282e+38 <op> 0.8 : != >
   [0.8 compare:3.40282e+38] : <
   [3.40282e+38 compare:0.8] : >
   [0.8 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqual:0.8] : NO
   [0.8 isEqualToNumber:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:0.8] : NO
C: 0.8 <op> 3.40282e+38 : < !=
C: 3.40282e+38 <op> 0.8 : != >
   [0.8 compare:3.40282e+38] : <
   [3.40282e+38 compare:0.8] : >
   [0.8 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqual:0.8] : NO
   [0.8 isEqualToNumber:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:0.8] : NO
C: 0.8 <op> 1.2e-63 : != >
C: 1.2e-63 <op> 0.8 : < !=
   [0.8 compare:1.2e-63] : >
   [1.2e-63 compare:0.8] : <
   [0.8 isEqual:1.2e-63] : NO
   [1.2e-63 isEqual:0.8] : NO
   [0.8 isEqualToNumber:1.2e-63] : NO
   [1.2e-63 isEqualToNumber:0.8] : NO
C: 0.8 <op> 1.2e+63 : < !=
C: 1.2e+63 <op> 0.8 : != >
   [0.8 compare:1.2e+63] : <
   [1.2e+63 compare:0.8] : >
   [0.8 isEqual:1.2e+63] : NO
   [1.2e+63 isEqual:0.8] : NO
   [0.8 isEqualToNumber:1.2e+63] : NO
   [1.2e+63 isEqualToNumber:0.8] : NO
C: 12.2 <op> 0.7 : != >
C: 0.7 <op> 12.2 : < !=
   [12.2 compare:0.7] : >
   [0.7 compare:12.2] : <
   [12.2 isEqual:0.7] : NO
   [0.7 isEqual:12.2] : NO
   [12.2 isEqualToNumber:0.7] : NO
   [0.7 isEqualToNumber:12.2] : NO
C: 12.2 <op> 1848 : < !=
C: 1848 <op> 12.2 : != >
   [12.2 compare:1848] : <
   [1848 compare:12.2] : >
   [12.2 isEqual:1848] : NO
   [1848 isEqual:12.2] : NO
   [12.2 isEqualToNumber:1848] : NO
   [1848 isEqualToNumber:12.2] : NO
C: 12.2 <op> 1.17549e-38 : != >
C: 1.17549e-38 <op> 12.2 : < !=
   [12.2 compare:1.17549e-38] : >
   [1.17549e-38 compare:12.2] : <
   [12.2 isEqual:1.17549e-38] : NO
   [1.17549e-38 isEqual:12.2] : NO
   [12.2 isEqualToNumber:1.17549e-38] : NO
   [1.17549e-38 isEqualToNumber:12.2] : NO
C: 12.2 <op> -1e-05 : != >
C: -1e-05 <op> 12.2 : < !=
   [12.2 compare:-1e-05] : >
   [-1e-05 compare:12.2] : <
   [12.2 isEqual:-1e-05] : NO
   [-1e-05 isEqual:12.2] : NO
   [12.2 isEqualToNumber:-1e-05] : NO
   [-1e-05 isEqualToNumber:12.2] : NO
C: 12.2 <op> 3.40282e+38 : < !=
C: 3.40282e+38 <op> 12.2 : != >
   [12.2 compare:3.40282e+38] : <
   [3.40282e+38 compare:12.2] : >
   [12.2 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqual:12.2] : NO
   [12.2 isEqualToNumber:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:12.2] : NO
C: 12.2 <op> 3.40282e+38 : < !=
C: 3.40282e+38 <op> 12.2 : != >
   [12.2 compare:3.40282e+38] : <
   [3.40282e+38 compare:12.2] : >
   [12.2 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqual:12.2] : NO
   [12.2 isEqualToNumber:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:12.2] : NO
C: 12.2 <op> 1.2e-63 : != >
C: 1.2e-63 <op> 12.2 : < !=
   [12.2 compare:1.2e-63] : >
   [1.2e-63 compare:12.2] : <
   [12.2 isEqual:1.2e-63] : NO
   [1.2e-63 isEqual:12.2] : NO
   [12.2 isEqualToNumber:1.2e-63] : NO
   [1.2e-63 isEqualToNumber:12.2] : NO
C: 12.2 <op> 1.2e+63 : < !=
C: 1.2e+63 <op> 12.2 : != >
   [12.2 compare:1.2e+63] : <
   [1.2e+63 compare:12.2] : >
   [12.2 isEqual:1.2e+63] : NO
   [1.2e+63 isEqual:12.2] : NO
   [12.2 isEqualToNumber:1.2e+63] : NO
   [1.2e+63 isEqualToNumber:12.2] : NO
C: 0.7 <op> 1848 : < !=
C: 1848 <op> 0.7 : != >
   [0.7 compare:1848] : <
   [1848 compare:0.7] : >
   [0.7 isEqual:1848] : NO
   [1848 isEqual:0.7] : NO
   [0.7 isEqualToNumber:1848] : NO
   [1848 isEqualToNumber:0.7] : NO
C: 0.7 <op> 1.17549e-38 : != >
C: 1.17549e-38 <op> 0.7 : < !=
   [0.7 compare:1.17549e-38] : >
   [1.17549e-38 compare:0.7] : <
   [0.7 isEqual:1.17549e-38] : NO
   [1.17549e-38 isEqual:0.7] : NO
   [0.7 isEqualToNumber:1.17549e-38] : NO
   [1.17549e-38 isEqualToNumber:0.7] : NO
C: 0.7 <op> -1e-05 : != >
C: -1e-05 <op> 0.7 : < !=
   [0.7 compare:-1e-05] : >
   [-1e-05 compare:0.7] : <
   [0.7 isEqual:-1e-05] : NO
   [-1e-05 isEqual:0.7] : NO
   [0.7 isEqualToNumber:-1e-05] : NO
   [-1e-05 isEqualToNumber:0.7] : NO
C: 0.7 <op> 3.40282e+38 : < !=
C: 3.40282e+38 <op> 0.7 : != >
   [0.7 compare:3.40282e+38] : <
   [3.40282e+38 compare:0.7] : >
   [0.7 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqual:0.7] : NO
   [0.7 isEqualToNumber:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:0.7] : NO
C: 0.7 <op> 3.40282e+38 : < !=
C: 3.40282e+38 <op> 0.7 : != >
   [0.7 compare:3.40282e+38] : <
   [3.40282e+38 compare:0.7] : >
   [0.7 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqual:0.7] : NO
   [0.7 isEqualToNumber:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:0.7] : NO
C: 0.7 <op> 1.2e-63 : != >
C: 1.2e-63 <op> 0.7 : < !=
   [0.7 compare:1.2e-63] : >
   [1.2e-63 compare:0.7] : <
   [0.7 isEqual:1.2e-63] : NO
   [1.2e-63 isEqual:0.7] : NO
   [0.7 isEqualToNumber:1.2e-63] : NO
   [1.2e-63 isEqualToNumber:0.7] : NO
C: 0.7 <op> 1.2e+63 : < !=
C: 1.2e+63 <op> 0.7 : != >
   [0.7 compare:1.2e+63] : <
   [1.2e+63 compare:0.7] : >
   [0.7 isEqual:1.2e+63] : NO
   [1.2e+63 isEqual:0.7] : NO
   [0.7 isEqualToNumber:1.2e+63] : NO
   [1.2e+63 isEqualToNumber:0.7] : NO
C: 1848 <op> 1.17549e-38 : != >
C: 1.17549e-38 <op> 1848 : < !=
   [1848 compare:1.17549e-38] : >
   [1.17549e-38 compare:1848] : <
   [1848 isEqual:1.17549e-38] : NO
   [1.17549e-38 isEqual:1848] : NO
   [1848 isEqualToNumber:1.17549e-38] : NO
   [1.17549e-38 isEqualToNumber:1848] : NO
C: 1848 <op> -1e-05 : != >
C: -1e-05 <op> 1848 : < !=
   [1848 compare:-1e-05] : >
   [-1e-05 compare:1848] : <
   [1848 isEqual:-1e-05] : NO
   [-1e-05 isEqual:1848] : NO
   [1848 isEqualToNumber:-1e-05] : NO
   [-1e-05 isEqualToNumber:1848] : NO
C: 1848 <op> 3.40282e+38 : < !=
C: 3.40282e+38 <op> 1848 : != >
   [1848 compare:3.40282e+38] : <
   [3.40282e+38 compare:1848] : >
   [1848 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqual:1848] : NO
   [1848 isEqualToNumber:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:1848] : NO
C: 1848 <op> 3.40282e+38 : < !=
C: 3.40282e+38 <op> 1848 : != >
   [1848 compare:3.40282e+38] : <
   [3.40282e+38 compare:1848] : >
   [1848 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqual:1848] : NO
   [1848 isEqualToNumber:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:1848] : NO
C: 1848 <op> 1.2e-63 : != >
C: 1.2e-63 <op> 1848 : < !=
   [1848 compare:1.2e-63] : >
   [1.2e-63 compare:1848] : <
   [1848 isEqual:1.2e-63] : NO
   [1.2e-63 isEqual:1848] : NO
   [1848 isEqualToNumber:1.2e-63] : NO
   [1.2e-63 isEqualToNumber:1848] : NO
C: 1848 <op> 1.2e+63 : < !=
C: 1.2e+63 <op> 1848 : != >
   [1848 compare:1.2e+63] : <
   [1.2e+63 compare:1848] : >
   [1848 isEqual:1.2e+63] : NO
   [1.2e+63 isEqual:1848] : NO
   [1848 isEqualToNumber:1.2e+63] : NO
   [1.2e+63 isEqualToNumber:1848] : NO
C: 1.17549e-38 <op> -1e-05 : != >
C: -1e-05 <op> 1.17549e-38 : < !=
   [1.17549e-38 compare:-1e-05] : >
   [-1e-05 compare:1.17549e-38] : <
   [1.17549e-38 isEqual:-1e-05] : NO
   [-1e-05 isEqual:1.17549e-38] : NO
   [1.17549e-38 isEqualToNumber:-1e-05] : NO
   [-1e-05 isEqualToNumber:1.17549e-38] : NO
C: 1.17549e-38 <op> 3.40282e+38 : < !=
C: 3.40282e+38 <op> 1.17549e-38 : != >
   [1.17549e-38 compare:3.40282e+38] : <
   [3.40282e+38 compare:1.17549e-38] : >
   [1.17549e-38 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqual:1.17549e-38] : NO
   [1.17549e-38 isEqualToNumber:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:1.17549e-38] : NO
C: 1.17549e-38 <op> 3.40282e+38 : < !=
C: 3.40282e+38 <op> 1.17549e-38 : != >
   [1.17549e-38 compare:3.40282e+38] : <
   [3.40282e+38 compare:1.17549e-38] : >
   [1.17549e-38 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqual:1.17549e-38] : NO
   [1.17549e-38 isEqualToNumber:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:1.17549e-38] : NO
C: 1.17549e-38 <op> 1.2e-63 : != >
C: 1.2e-63 <op> 1.17549e-38 : < !=
   [1.17549e-38 compare:1.2e-63] : >
   [1.2e-63 compare:1.17549e-38] : <
   [1.17549e-38 isEqual:1.2e-63] : NO
   [1.2e-63 isEqual:1.17549e-38] : NO
   [1.17549e-38 isEqualToNumber:1.2e-63] : NO
   [1.2e-63 isEqualToNumber:1.17549e-38] : NO
C: 1.17549e-38 <op> 1.2e+63 : < !=
C: 1.2e+63 <op> 1.17549e-38 : != >
   [1.17549e-38 compare:1.2e+63] : <
   [1.2e+63 compare:1.17549e-38] : >
   [1.17549e-38 isEqual:1.2e+63] : NO
   [1.2e+63 isEqual:1.17549e-38] : NO
   [1.17549e-38 isEqualToNumber:1.2e+63] : NO
   [1.2e+63 isEqualToNumber:1.17549e-38] : NO
C: -1e-05 <op> 3.40282e+38 : < !=
C: 3.40282e+38 <op> -1e-05 : != >
   [-1e-05 compare:3.40282e+38] : <
   [3.40282e+38 compare:-1e-05] : >
   [-1e-05 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqual:-1e-05] : NO
   [-1e-05 isEqualToNumber:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:-1e-05] : NO
C: -1e-05 <op> 3.40282e+38 : < !=
C: 3.40282e+38 <op> -1e-05 : != >
   [-1e-05 compare:3.40282e+38] : <
   [3.40282e+38 compare:-1e-05] : >
   [-1e-05 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqual:-1e-05] : NO
   [-1e-05 isEqualToNumber:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:-1e-05] : NO
C: -1e-05 <op> 1.2e-63 : < !=
C: 1.2e-63 <op> -1e-05 : != >
   [-1e-05 compare:1.2e-63] : <
   [1.2e-63 compare:-1e-05] : >
   [-1e-05 isEqual:1.2e-63] : NO
   [1.2e-63 isEqual:-1e-05] : NO
   [-1e-05 isEqualToNumber:1.2e-63] : NO
   [1.2e-63 isEqualToNumber:-1e-05] : NO
C: -1e-05 <op> 1.2e+63 : < !=
C: 1.2e+63 <op> -1e-05 : != >
   [-1e-05 compare:1.2e+63] : <
   [1.2e+63 compare:-1e-05] : >
   [-1e-05 isEqual:1.2e+63] : NO
   [1.2e+63 isEqual:-1e-05] : NO
   [-1e-05 isEqualToNumber:1.2e+63] : NO
   [1.2e+63 isEqualToNumber:-1e-05] : NO
C: 3.40282e+38 <op> 3.40282e+38 : ==
C: 3.40282e+38 <op> 3.40282e+38 : ==
   [3.40282e+38 compare:3.40282e+38] : =
   [3.40282e+38 compare:3.40282e+38] : =
   [3.40282e+38 isEqual:3.40282e+38] : YES
   [3.40282e+38 isEqual:3.40282e+38] : YES
   [3.40282e+38 isEqualToNumber:3.40282e+38] : YES
   [3.40282e+38 isEqualToNumber:3.40282e+38] : YES
C: 3.40282e+38 <op> 1.2e-63 : != >
C: 1.2e-63 <op> 3.40282e+38 : < !=
   [3.40282e+38 compare:1.2e-63] : >
   [1.2e-63 compare:3.40282e+38] : <
   [3.40282e+38 isEqual:1.2e-63] : NO
   [1.2e-63 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:1.2e-63] : NO
   [1.2e-63 isEqualToNumber:3.40282e+38] : NO
C: 3.40282e+38 <op> 1.2e+63 : < !=
C: 1.2e+63 <op> 3.40282e+38 : != >
   [3.40282e+38 compare:1.2e+63] : <
   [1.2e+63 compare:3.40282e+38] : >
   [3.40282e+38 isEqual:1.2e+63] : NO
   [1.2e+63 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:1.2e+63] : NO
   [1.2e+63 isEqualToNumber:3.40282e+38] : NO
C: 3.40282e+38 <op> 1.2e-63 : != >
C: 1.2e-63 <op> 3.40282e+38 : < !=
   [3.40282e+38 compare:1.2e-63] : >
   [1.2e-63 compare:3.40282e+38] : <
   [3.40282e+38 isEqual:1.2e-63] : NO
   [1.2e-63 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:1.2e-63] : NO
   [1.2e-63 isEqualToNumber:3.40282e+38] : NO
C: 3.40282e+38 <op> 1.2e+63 : < !=
C: 1.2e+63 <op> 3.40282e+38 : != >
   [3.40282e+38 compare:1.2e+63] : <
   [1.2e+63 compare:3.40282e+38] : >
   [3.40282e+38 isEqual:1.2e+63] : NO
   [1.2e+63 isEqual:3.40282e+38] : NO
   [3.40282e+38 isEqualToNumber:1.2e+63] : NO
   [1.2e+63 isEqualToNumber:3.40282e+38] : NO
C: 1.2e-63 <op> 1.2e+63 : < !=
C: 1.2e+63 <op> 1.2e-63 : != >
   [1.2e-63 compare:1.2e+63] : <
   [1.2e+63 compare:1.2e-63] : >
   [1.2e-63 isEqual:1.2e+63] : NO
   [1.2e+63 isEqual:1.2e-63] : NO
   [1.2e-63 isEqualToNumber:1.2e+63] : NO
   [1.2e+63 isEqualToNumber:1.2e-63] : NO
